### PR TITLE
Clarify Type System and Operator Behavior

### LINKED ISSUE #40: Do commas have a syntactical meaning?
Some of the example loops have a comma at the end of them:

```
Tommy was a dancer
While Tommy ain't nothing,
Knock Tommy down
```

but others don't:
```
Tommy was a dancer
While Tommy ain't nothing
Shout it
Knock it down

```

Are these commas supposed to be here? If they are, do they do anything or are they just ignored?

### LINKED ISSUE #97: Weak typing?
It would be a good idea to clarify if Rockstar is weakly or strongly typed.

Specifically:
- can I read a string from stdin and silently convert it into a number?
- can I add a string and a number, obtaining the original string concatenated to the string representation of the number?

For example, can I do this?
```Listen to your words
Build your words up
The answer is "I love "
Put the answer with your words into my voice
Shout my voice 
```
stdin: "5"
stdout: "I love 6"

### LINKED ISSUE #101: Add spec for boolean operators (and/or)
They're being used in the examples, but they're not in the spec yet.

Operator precedence should also be specified. Based on the examples, `and` definitely binds outside of function calls, but I can't really tell beyond that.

### LINKED ISSUE #102: Function syntax for functions with 0 arguments is not defined

### LINKED ISSUE #107: Truthy values
The "Conditionals" section says `If the expression evaluates to true` but doesn't define what's "true". In many languages, there's a "truthy" set of values and everything else is "false" (or vice-versa depending on the whims of the language designers). Any thoughts?

(Ran into this in #105 which worked because in the Python interpreter "0" was treated as a boolean value)

### LINKED ISSUE #108: String concatenation
cellular-rocktomata (#105) relies on being able to do string concatenation with the `plus`/`with` operator, which currently isn't part of the spec, but I'm not seeing any other way to do this currently. Can this be added? Similarly `times` with "&lt;string&gt; `times` &lt;integer&gt;" might be of use.

### LINKED ISSUE #130: Do function calls allow "and" for separating arguments?
When describing function calls, the spec doesn't explicitly say what can separate arguments, but both examples given use commas, such as `Multiply taking 3, 5`.  However, the minimalist (non-idiomatic) fizzbuzz code includes `If Modulus taking Counter and Fizz is 0` which seems to call the Modulus function with Counter and Fizz as arguments. This seems like it creates an ambiguity since the line could be interpreted as `if Modulus(Counter, Fizz) == 0` and `if Modulus(Counter) && Fizz == 0`.

To get to my actual question, is it intentional that "and" is used as an argument separator in function calls in one of the examples, or does Rockstar only support using commas to separate arguments?

### PR DESCRIPTION
**Pull request by cwfitzgerald**
This is a doozy. 

Resolves #40 (Commas in function calls)
Resolves #97 (Weak Typing)
Resolves #101 (Boolean Operators)
Resolves #102 (Functions with zero arguments)
Resolves #107 (Truthiness)
Resolves #108 (String Concatenation)
Resolves #130 (Function calls with `and`)

This tries to stay as accurate as possible, and I've looked over everything thrice, but I have probably missed an edge case or the wording isn't as precise as it could be. I have tried to keep to what was agreed upon in various discussions, but I will imagine some lively discussion will come of this. 🤘

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -192,6 +214,64 @@ Use the `Say` keyword to write the value of a variable to `STDOUT`.
...
+6. `and`, `or`, and `nor` (left-associative)
+
+##### Examples
+
+- `A taking B times C plus not D times E and F` is equivalent to `((A(B) * C) + (!D * E)) && F`
+- `A taking B taking C, D` is equivalent to `A(B(C, D)` **NOT** `A(B(C), D)`.
+
+#### Binary Comparison
+
+Equality comparisons (`is`, `ain't`, `is not`) are allowed between any two operands of the same type. Objects are checked by reference equality, all other types are checked by value equality.
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply README.md]: yeah, that really should have the same coercion spec as ordered operators.

**Comment by wizered67** [comment]
Not a huge deal, but it does bother me a little that `A is not B` has two interpretations that behave differently. I know the spec clarifies that it is always `A != B` but it still feels a bit weird to me. What if instead of using "is not" for comparison we use "isn't" instead, in addition to the existing alternative of "aint", so that `A is not B` only has a single interpretation. Not set on this though if you prefer keeping "is not" as is.

**Comment by wizered67** [comment]
Just realized, it would be good to clarify if the logical operators are "short circuiting". For example, is `false and 1 over 0` equal to `false` or an error?

**Comment by cwfitzgerald** [comment]
I agree on both counts, I'll give it a fix. Operators should short circuit, they wouldn't be very useful otherwise.

**Comment by wizered67** [comment]
Thanks @cwfitzgerald! I believe there's a typo in your new commit - it says `true and 1 over 0` is `false` when it should be `false and 1 over 0`.

**Comment by cwfitzgerald** [comment]
@wizered67 and that my friends is why peer review exists. That's what I get for editing while tired. :)

**Comment by wizered67** [comment]
Looking over this again, I'm realizing there doesn't seem to be a way to read a string from standard input and convert it to a number. Using it in a comparison would convert it to a number, but trying to do math on it would treat it as a string.

**Comment by marcinruszkiewicz** [comment]
I convert it implicitly in Kaiser on reading from stdin, out of the assumption that if someone enters a number they would most likely expect it to be treated like a number.

Some conversion function might be nice though.

**Comment by cwfitzgerald** [comment]
@marcinruszkiewicz This problem is going to be solved by the modules/stdlib implementation that have been mused about in various places. #115 is the place that my github-foo can find.

**Comment by wizered67** [comment]
A standard library seems like a good idea! It would be nice to start standardizing it soon before every implementation makes their own functions.

**Comment by cwfitzgerald** [comment]
A agree, it's on my todo list 😄

File: README.md [hunk 2]
```diff
@@ -192,6 +216,64 @@ Use the `Say` keyword to write the value of a variable to `STDOUT`.
...
+
+- `A taking B times C plus not D times E and F` is equivalent to `((A(B) * C) + (!D * E)) && F`
+
+#### Binary Comparison
+
+Equality comparisons (`is`, `ain't`, `is not`) are allowed between types if they are the same type or they can be compared by the rules below. Objects are checked by reference equality, all other types are checked by value equality.
+
+Ordering comparisons (`is higher than`, `is lower than`, `is as high as`, and `is as low as`) are only allowed if the operands are both Numbers or both Strings or they are converted to such an arrangement according to the rules below. Numbers are compared as expected, Strings are compared lexicographically.
+
+- \<Mysterious\> \<op\> Mysterious =\> Equal.
```
Comment (marcinruszkiewicz) [inline README.md]: Probably doesn't need the `< >` if you're not gonna use them on the right side of the operation (same below with Non-Mysterious

**Pull request merged**